### PR TITLE
Expose RunData fields (metrics, params, tags) as dictionaries in Python API

### PR DESCRIPTION
**Pull request by smurching**
The changes in this PR are summarized below. Note that the PR makes no protobuf (REST API) changes, so the only user-visible changes are in how run data is accessed in the Python API. The PR is tested by updating existing store & fluent API unit tests, which verify that well-formed `RunData` instances are constructed & returned along the read path (note that `RunData` is not used in the write path). See [this gist](https://gist.github.com/smurching/f60f7c418ec03f2bcee0f3374bfa1d81) for additional context & discussion of design decisions.

#### Breaking change: RunData fields are now dicts
In the public Python API: RunData fields (`metrics`, `params`, `tags`) are now dictionaries instead of lists:
    * `metrics` is a dictionary of metric key (string) to metric value.
    * `params` is a dictionary of param key (string) to param value.
    * `tags` is a dictionary of tag key (string) to tag value.

Existing code written to access metrics/params/tags by key, e.g:

```
run = mlflow.get_run("925dbee9028f415fb66d1005a9d90f2f")
tag_obj = [tag for tag in run.data.tags if tag.key == "mlflow.docker"][0]
metric_obj = [metric for metric in run.data.metrics if metric.key == "rmse"][0]
param_obj = [param for param in run.data.params if param.key == "alpha"][0]
print("Tag key: %s, value: %s" % (tag.key, tag.value))
print("Metric key: %s, value: %s, timestamp: %s" % (metric.key, metric.value, metric.timestamp))
print("Param key: %s, value: %s" % (param_obj.key, param_obj.value))
```

can now be written as:

```
run = mlflow.get_run("925dbee9028f415fb66d1005a9d90f2f")
tag_val = run.data.tags["mlflow.docker"] # tag_val is a string, e.g. 'true'
metric_val = run.data.metrics["rmse"] # metric_val is a float, e.g. 0.1
param_val = run.data.params["alpha"] # param_value is a string, e.g. '1.0'
print("Tag key: %s, value: %s" % ("mlflow.docker", tag_val))
print("Metric value: %s" % ("rmse", metric_val))
print("Param key: %s, value: %s" % ("alpha", param_val))
```

Another example of querying a run & its data:
```
>>> run = mlflow.tracking.MlflowClient().get_run("252b2fc8342b49239a3f819988e37994")
>>> run
<Run: data=<RunData: metrics={'mae': 0.6177463381374035,
 'r2': 0.19042414117747397,
 'rmse': 0.7917149365733894}, params={'alpha': '0.42', 'l1_ratio': '0.1'}, tags={'mlflow.gitRepoURL': 'https://github.com/mlflow/mlflow'}>, info=<RunInfo: artifact_uri='/Users/siddharthmurching/code/mlflow/mlruns/10/252b2fc8342b49239a3f819988e37994/artifacts', end_time=1554499565539, entry_point_name='main', experiment_id=10, lifecycle_stage='active', name='', run_uuid='252b2fc8342b49239a3f819988e37994', source_name='file:///Users/siddharthmurching/code/mlflow#examples/sklearn_elasticnet_wine', source_type=3, source_version='2d5aacf8ef103469334c4f3945b2a28634797b13', start_time=1554499562896, status=3, user_id='siddharthmurching'>>
>>> run.data.metrics['mae']
0.6177463381374035
>>> run.data.params['alpha']
'0.42'
>>> run.data.tags['mlflow.gitRepoURL']
'https://github.com/mlflow/mlflow'
```

#### RunData constructor remains unchanged
Note that the `mlflow.entities.RunData` constructor still accepts lists of metric/param/tag entities, to make it simple to construct a RunData entity from its protobuf equivalent & allow for exposing proto fields (like lists of metric / param / tag objects) via RunData in the future if desired

### CODE REVIEW

File: mlflow/entities/_mlflow_object.py [hunk 1]
```diff
@@ -44,12 +43,6 @@ def __init__(self):
     def to_string(self, obj):
         if isinstance(obj, _MLflowObject):
             return "<%s: %s>" % (get_classname(obj), self._entity_to_string(obj))
-        # Handle nested lists inside MLflow entities (e.g. lists of metrics/params)
```
Comment (smurching) [inline mlflow/entities/_mlflow_object.py]: Removed this code path as we no longer have lists nested inside of MLflow entities (we have dicts instead). Note that this means we don't truncate the string representations of runs with many distinct metrics/params/tags. I could try to add back such truncation logic, but it's not obvious to me that it'd be super useful.

File: mlflow/store/file_store.py [hunk 2]
```diff
@@ -424,9 +424,10 @@ def _get_metric_from_file(parent_path, metric_name):
     def get_all_metrics(self, run_uuid):
         _validate_run_id(run_uuid)
         parent_path, metric_files = self._get_run_files(run_uuid, "metric")
-        metrics = []
+        metrics = {}
```
Comment (smurching) [inline mlflow/store/file_store.py]: These functions are all called into by `get_run` (which is also their only caller outside of tests)

File: tests/entities/test_run_data.py [hunk 3]
```diff
@@ -1,30 +1,34 @@
...
 class TestRunData(unittest.TestCase):
     def _check_metrics(self, metrics_1, metrics_2):
-        for metric in metrics_1:
+        for key, metric in metrics_1.items():
             self.assertIsInstance(metric, Metric)
-        self.assertEqual(set([m.key for m in metrics_1]), set([m.key for m in metrics_2]))
-        self.assertEqual(set([m.value for m in metrics_1]), set([m.value for m in metrics_2]))
-        self.assertEqual(set([m.timestamp for m in metrics_1]),
-                         set([m.timestamp for m in metrics_2]))
+            self.assertEqual(metric.key, key)
```
Comment (smurching) [inline tests/entities/test_run_data.py]: As a sanity check, make sure the metric key matches the dict key

File: tests/entities/test_run.py [hunk 4]
```diff
@@ -53,14 +53,15 @@ def test_string_repr(self):
...
         run1 = Run(run_info, run_data)
-        expected = "<Run: data=<RunData: metrics=[<Metric: key='key', timestamp=0, value=0>, " \
-                   "<Metric: key='key', timestamp=0, value=1>, ...], params=[], tags=[]>, " \
-                   "info=<RunInfo: artifact_uri=None, end_time=1, " \
-                   "entry_point_name='entry-point-name', experiment_id=0, " \
-                   "lifecycle_stage='active', name='name', run_uuid='hi', " \
-                   "source_name='source-name', source_type=3, source_version='version', " \
-                   "start_time=0, status=4, user_id='user-id'>>"
+        expected = ("<Run: data=<RunData: metrics={'key-0': <Metric: key='key-0', timestamp=0, "
+                    "value=0>,\n 'key-1': <Metric: key='key-1', timestamp=0, value=1>,\n "
```
Comment (smurching) [inline tests/entities/test_run.py]: It looks like Python's pprint library adds these newlines between successive key-value pairs in the same dictionary, which I guess improves readability a bit.

File: mlflow/utils/search_utils.py [hunk 5]
```diff
@@ -241,22 +241,21 @@ def does_run_match_clause(cls, run, sed):
                                       "not one of '%s" % (comparator,
                                                           cls.VALID_METRIC_COMPARATORS),
                                       error_code=INVALID_PARAMETER_VALUE)
-            metric = next((m for m in run.data.metrics if m.key == key), None)
+            metric = next((m for m in run.data.metrics.values() if m.key == key), None)
```
Comment (tomasatdatabricks) [inline mlflow/utils/search_utils.py]: why not:
``` 
run.data.metrics[m.key]  if m.key in run.data.metrics else None
```
?

File: mlflow/utils/search_utils.py [hunk 6]
```diff
@@ -241,22 +241,21 @@ def does_run_match_clause(cls, run, sed):
...
             lhs = metric.value if metric else None
             value = float(value)
         elif key_type == cls._PARAM_IDENTIFIER:
             if comparator not in cls.VALID_PARAM_COMPARATORS:
                 raise MlflowException("Invalid comparator '%s' "
                                       "not one of '%s" % (comparator, cls.VALID_PARAM_COMPARATORS),
                                       error_code=INVALID_PARAMETER_VALUE)
-            param = next((p for p in run.data.params if p.key == key), None)
-            lhs = param.value if param else None
+            lhs = next((p_value for p_key, p_value in run.data.params.items() if p_key == key),
```
Comment (tomasatdatabricks) [inline mlflow/utils/search_utils.py]: same as above

File: mlflow/utils/search_utils.py [hunk 7]
```diff
@@ -241,22 +241,21 @@ def does_run_match_clause(cls, run, sed):
...
+                       None)
         elif key_type == cls._TAG_IDENTIFIER:
             if comparator not in cls.VALID_TAG_COMPARATORS:
                 raise MlflowException("Invalid comparator '%s' "
                                       "not one of '%s" % (comparator, cls.VALID_TAG_COMPARATORS))
-            tag = next((t for t in run.data.tags if t.key == key), None)
-            lhs = tag.value if tag else None
+            lhs = next((t_value for t_key, t_value in run.data.tags.items() if t_key == key), None)
```
Comment (tomasatdatabricks) [inline mlflow/utils/search_utils.py]: same as ^^

File: tests/entities/test_run.py [hunk 8]
```diff
@@ -35,9 +35,9 @@ def test_creation_and_hydration(self):
                             "lifecycle_stage": lifecycle_stage,
                             "artifact_uri": artifact_uri,
                             },
-                   "data": {"metrics": [dict(m) for m in metrics],
-                            "params": [dict(p) for p in params],
-                            "tags": [dict(t) for t in tags]}}
+                   "data": {"metrics": {m.key: dict(m) for m in metrics.values()},
```
Comment (tomasatdatabricks) [inline tests/entities/test_run.py]: This is unrelated to this PR, but I wonder, do we call to_dictionary outside of tests? Should we just compare directly without the indirection?

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good, thx Sid! I just had a few questions.

File: .travis.yml [hunk 9]
```diff
@@ -82,7 +82,7 @@ script:
   - which mlflow
   - echo $MLFLOW_HOME
   - SAGEMAKER_OUT=$(mktemp)
-  - if mlflow sagemaker build-and-push-container --no-push --mlflow-home . > $SAGEMAKER_OUT 2>&1; then
```
[the remark this answers is not in the document]
Comment (smurching) [reply .travis.yml]: Oops, this was an unnecessary attempt to debug Sagemaker test failures unrelated to this PR. will remove!

File: mlflow/entities/run_data.py [hunk 10]
```diff
@@ -2,66 +2,71 @@
...
         if isinstance(metric, dict):
             metric = Metric(metric['key'], metric['value'], metric['timestamp'])
-        self._metrics.append(metric)
+        self._metrics[metric.key] = metric
 
     def _add_param(self, param):
         if isinstance(param, dict):
             param = Param(param['key'], param['value'])
-        self._params.append(param)
+        self._params[param.key] = param.value
```
[the remark this answers is not in the document]
Comment (smurching) [reply mlflow/entities/run_data.py]: Good question, it's because metrics have additional fields besides their value (like timestamp, and in the near future, x coordinates)

File: mlflow/utils/search_utils.py (hunk 5 above)
Comment (smurching) [reply mlflow/utils/search_utils.py]: Your suggestion seems clearer to me, will make that change :)

File: mlflow/entities/run_data.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tomasatdatabricks) [reply mlflow/entities/run_data.py]: I guess I was more thinking if it would make sense to keep the other ones as map to Param / Tag instead of directly pointing to the value for the sake of consistency.

[the remark this answers is not in the document]
Comment (smurching) [reply mlflow/entities/run_data.py]: Yeah I think that's a fair point, this is actually a design point that probably should have been discussed more broadly before starting implementation. I guess the questions are:

1) Are there significant ease-of-use gains from being able to access tag/param values within a flat dictionary (e.g. `param_val = run.data.params["alpha"]`) vs having to perform an extra access to get to the value (`param_val = run.data.params["alpha"].value`)? If so, are these ease-of-use gains worth a) the inconsistency of having metrics be a dict of string -> Metric object, and b) the inflexibility of not being able to add new fields to params/tags without breaking API changes?

File: tests/entities/test_run.py (hunk 8 above)
Comment (smurching) [reply tests/entities/test_run.py]: Good question, converting to dict helps us bypass the fact that MLflow entities with the same properties aren't actually considered equal (https://github.com/mlflow/mlflow/pull/961/files is an attempt to address this). We only use `to_dictionary` in one place outside tests - in the runs CLI ([link](https://github.com/mlflow/mlflow/blob/f1629c2/mlflow/runs.py#L83)).

File: mlflow/entities/run_data.py [hunk 11]
```diff
@@ -2,66 +2,71 @@
...
         return self._tags
 
     def _add_metric(self, metric):
         if isinstance(metric, dict):
             metric = Metric(metric['key'], metric['value'], metric['timestamp'])
-        self._metrics.append(metric)
+        self._metrics[metric.key] = metric
 
     def _add_param(self, param):
         if isinstance(param, dict):
```
[the remark this answers is not in the document]
Comment (smurching) [reply mlflow/entities/run_data.py]: I think originally it was so we could call `_add_param` in both `from_dictionary` and `from_proto` and pass either a dictionary representation of a param or the Param object itself.

File: tests/entities/test_run.py (hunk 8 above)
Comment (smurching) [reply tests/entities/test_run.py]: Created https://github.com/mlflow/mlflow/issues/1089 to track removing the to_dictionary, from_dictionary APIs for MLflow 1.0

Comment (tomasatdatabricks) [reply tests/entities/test_run.py]: awesome, thanks Sid!

File: mlflow/entities/run_data.py [hunk 12]
```diff
@@ -2,66 +2,69 @@
...
+        self._metrics = {metric.key: metric.value for metric in self._metric_objs}
+        self._params = {param.key: param.value for param in (params or [])}
+        self._tags = {tag.key: tag.value for tag in (tags or [])}
 
     @property
     def metrics(self):
-        """List of :py:class:`mlflow.entities.Metric` for the current run."""
+        """
+        Dictionary of string key -> metric value for the current run.
+        For each metric key, the maximum metric value at the maximum timestamp is returned.
```
Comment (tomasatdatabricks) [inline mlflow/entities/run_data.py]: nit: I would maybe rephrase that to make it clear that we do not expect to have ties. e.g. something like : ... the metric value with the latest timestamp is returned. In case there are multiple values with the same latest timestamp, the maximum of these values is returned.

Just a suggestion though.

File: mlflow/entities/run_data.py [hunk 13]
```diff
@@ -2,66 +2,69 @@
...
-            tag = RunTag(tag['key'], tag['value'])
-        self._tags.append(tag)
+        self._tags[tag.key] = tag.value
 
     def to_proto(self):
         run_data = ProtoRunData()
-        run_data.metrics.extend([m.to_proto() for m in self.metrics])
-        run_data.params.extend([p.to_proto() for p in self.params])
-        run_data.tags.extend([t.to_proto() for t in self.tags])
+        run_data.metrics.extend([m.to_proto() for m in self._metric_objs])
```
Comment (tomasatdatabricks) [inline mlflow/entities/run_data.py]: nit: why extend? why not just assign the value?

**Review by tomasatdatabricks — APPROVED** [review]
lgtm! I left few nit comments.

File: mlflow/entities/run_data.py (hunk 13 above)
Comment (smurching) [reply mlflow/entities/run_data.py]: Apparently protobuf doesn't allow you to assign to `repeated` fields within a protobuf object (from Travis logs):

```
  File "/home/travis/build/mlflow/mlflow/mlflow/server/handlers.py", line 272, in _get_run
response_message.run.MergeFrom(_get_store().get_run(request_message.run_uuid).to_proto())
  File "/home/travis/build/mlflow/mlflow/mlflow/entities/run.py", line 40, in to_proto
    run.data.MergeFrom(self.data.to_proto())
  File "/home/travis/build/mlflow/mlflow/mlflow/entities/run_data.py", line 58, in to_proto
    run_data.metrics = [m.to_proto() for m in self._metric_objs]
AttributeError: Assignment not allowed to repeated field "metrics" in protocol message object.
```

**Comment by smurching** [comment]
Thanks @tomasatdatabricks! Merging to master :)
